cloud-controller-manager/do/resources.go

```diff
@@ -34,63 +35,228 @@ import (
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	c.loadBalancerIDMap = newIDMap
+	c.loadBalancerNameMap = newNameMap
+}
+
 // ResourcesController ensures that DO resources are properly tagged.
 type ResourcesController struct {
-	clusterIDTag string
-	kclient      kubernetes.Interface
-	gclient      *godo.Client
-	svcLister    v1lister.ServiceLister
+	clusterID string
+	kclient   kubernetes.Interface
+	gclient   *godo.Client
+	svcLister v1lister.ServiceLister
+
+	resources *resources
 }
 
-// NewResourcesController returns a new services controller responsible for
-// re-applying cluster IDs on DO load-balancers.
-func NewResourcesController(clusterIDTag string, inf v1informers.ServiceInformer,
-	k kubernetes.Interface, g *godo.Client) *ResourcesController {
+// NewResourcesController returns a new controller responsible for managing
+// DigitalOcean cloud resources.
+func NewResourcesController(
+	clusterID string,
+	r *resources,
+	inf v1informers.ServiceInformer,
+	k kubernetes.Interface,
+	g *godo.Client,
+) *ResourcesController {
 	return &ResourcesController{
-		clusterIDTag: clusterIDTag,
-		kclient:      k,
-		gclient:      g,
-		svcLister:    inf.Lister(),
+		clusterID: clusterID,
+		resources: r,
+		kclient:   k,
+		gclient:   g,
+		svcLister: inf.Lister(),
 	}
 }
 
 // Run starts the resources controller. It watches over DigitalOcean resources
 // making sure the right tags are set.
 func (r *ResourcesController) Run(stopCh <-chan struct{}) {
-	ticker := time.NewTicker(controllerSyncPeriod)
-	defer ticker.Stop()
-	// Do not wait for initial tick to pass; run immediately for reduced sync
-	// latency.
-	for tickerC := ticker.C; ; {
-		if err := r.sync(); err != nil {
-			glog.Errorf("failed to sync load-balancer tags: %s", err)
-		}
-		select {
-		case <-tickerC:
-			continue
-		case <-stopCh:
-			return
+	syncResourcesTicker := time.NewTicker(controllerSyncResourcesPeriod)
+	defer syncResourcesTicker.Stop()
+
+	go func() {
+		// Do not wait for initial tick to pass; run immediately for reduced sync
+		// latency.
+		for tickerC := syncResourcesTicker.C; ; {
+			r.syncResources()
+			select {
+			case <-tickerC:
+				continue
+			case <-stopCh:
+				return
+			}
 		}
+	}()
+
+	if r.clusterID == "" {
+		glog.Info("No cluster ID configured -- skipping tags syncing.")
+		return
 	}
+
+	syncTagsTicker := time.NewTicker(controllerSyncTagsPeriod)
+	defer syncTagsTicker.Stop()
+
+	go func() {
+		// Do not wait for initial tick to pass; run immediately for reduced sync
+		// latency.
+		for tickerC := syncTagsTicker.C; ; {
+			if err := r.syncTags(); err != nil {
+				glog.Errorf("failed to sync load-balancer tags: %s", err)
+			}
+			select {
+			case <-tickerC:
+				continue
+			case <-stopCh:
+				return
+			}
+		}
+	}()
 }
 
-func (r *ResourcesController) sync() error {
-	// Fetch all load-balancers.
-	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+func (r *ResourcesController) syncResources() {
```

Comment:
Can you add a doc comment that this function is responsible of updating the cache with up to date information?

Reply:
Done.